plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -149,4 +174,45 @@ private void generateSplits(boolean useIndex)
             throw new TrinoException(HUDI_CANNOT_OPEN_SPLIT, "Error generating Hudi split", e);
         }
     }
+
+    private Map<String, Partition> getPartitionInfos()
+    {
+        if (!enableMetadataTable || !isMetadataPartitionListingEnabled) {
+            return lazyPartitionMap.get();
+        }
+
+        try {
+            PartitionValueExtractor partitionValueExtractor = getPartitionValueExtractor(lazyMetaClient.get().getTableConfig());
+            List<HiveColumnHandle> hivePartitionColumns = tableHandle.getPartitionColumns();
+
+            List<Column> partitionColumns = hivePartitionColumns.stream()
+                    .map(column -> new Column(
+                            column.getName(),
+                            column.getHiveType(),
+                            column.getComment(),
+                            Collections.emptyMap()))
+                    .toList();
+            log.info("Listing partitions for %s.%s via metadata table using partition value extractor %s",
+                    tableHandle.getSchemaName(), tableHandle.getTableName(), partitionValueExtractor.getClass().getSimpleName());
+            Map<String, Partition> metadataPartitions = lazyTableMetadata.get()
+                    .getAllPartitionPaths().stream()
+                    .map(partitionPath -> HudiUtil.buildPartition(partitionPath, partitionColumns, tableHandle, partitionValueExtractor))
+                    .collect(Collectors.toMap(
+                            HudiUtil::getHivePartitionName,
+                            Function.identity()));
+
+            if (metadataPartitions.isEmpty()) {
+                log.warn("No partitions found via metadata table for %s.%s, switching to metastore-based listing.",
+                        tableHandle.getSchemaName(), tableHandle.getTableName());
+                return lazyPartitionMap.get();
+            }
```

Comment:
I'm thinking about the case where the table registered in the catalog has both Hudi and non-Hudi partitions.  This used to work in Hive connector support.  @voonhous I remember older Hudi release used to support this case for tables migrated from parquet, correct?  We can add a test on this case in a separate PR.

Reply:
Yes this is what I'm referring to.  Usually the table is date-partitioned and Hudi starts to write new partitions only:
```
date=20250701/ 
date=20250702/
date=20250703/
date=20250704/    <- written by parquet, old Hive partitions (non-Hudi partitions)
date=20250705/    <- written by Hudi, only new partitions are updated by Hudi
date=20250706/
date=20250707/
```
Previously Hudi provides the ability to query all non-Hudi and Hudi partitions.

We can address this separately as I'm not sure about the prevalence of such use case right now.